### PR TITLE
Add 'empty dungeons' setting

### PR DESCRIPTION
**Pull request by damdayou**
This setting adds the ability to select or randomly pick _empty dungeons_. _Empty dungeons_ are major dungeons filled with junk items and whose rewards are given to the player for free from the title screen. Simply put, these dungeons won't contain any useful item and won't have to be completed in order to beat the seed.

This setting offers two options:
- _Specific dungeons_: choose which specific major dungeons will be empty,
- _Count_: choose how many empty dungeons will be randomly chosen.

Compatibility with other main settings has been guaranteed as follows.
- **MQ Dungeons**: If empty dungeons and MQ dungeons are both on, the generator ensures (whenever it's possible) that a same major dungeon is not both MQ and empty (whenever it's possible).
- **Shuffle dungeon entrances**: If this setting is on, the 'empty' property refers to the dungeon itself, not the entrance. _Note that when Spirit Temple is empty, one might still need to visit it to access Desert Colossus._
- **Shuffle boss**: The proper boss room is emptied and the proper reward is given for free. E.g. if Forest Temple is empty and Phantom Ganon has been replaced by Volvagia, then there is a junk item on Volvagia for sure but there may be a major item on Phantom Ganon (if he is in a non-empty dungeons).
- **Dungeons Have One Major Item**: This setting has to be interpreted as _Non-empty Dungeons Have One Major Item_. Empty dungeons won't have a major item in any case.
- **Shuffle Songs**: If this setting is set to _Dungeon Rewards_, then the songs that would have been placed in empty dungeons will be given to the player for free from the title screen. 
- **Shuffle Map/Compass/Small Keys/Boss key**: These settings will only affect dungeon items from non-empty dungeons. Unless removed, dungeon items from empty dungeons will always remain in their own dungeons.

### CODE REVIEW

**Comment by damdayou** [comment]
Yes sorry, I did a mistake and merge the opposite way I wanted to. I'm currently trying to revert this commit, sorry for the inconvenience.

**Comment by damdayou** [comment]
I tried to update the tooltips following the comments in https://discord.com/channels/274180765816848384/512048482677424138/987763948218626099

Feel free to tell me if my grammar is incorrect or the phrasing is awkward, English is not my native language so I would be surprised there isn't a single mistake.

File: SettingsList.py [hunk 1]
```diff
@@ -4041,6 +4054,83 @@ def __init__(self, name, gui_text, min, max, default, step=1,
...
+        name           = 'empty_dungeons_mode',
+        gui_text       = 'Empty dungeons',
+        default        = 'none',
+        choices        = {
+            'none':       'Off',
+            'specific':   'Specific Dungeons',
+            'count':      'Count',
+        },
+        gui_tooltip    = '''\
+            Empty dungeons are dungeons guaranted to be barren, and whose rewards
```
Comment (ETR-BTF) [inline SettingsList.py:4067]: For the sake of clarity, I think it'd be better to explicitly call them "Dungeon Rewards" instead of "rewards".

"guaranteed" is with two e's

File: SettingsList.py [hunk 2]
```diff
@@ -4041,6 +4054,83 @@ def __init__(self, name, gui_text, min, max, default, step=1,
...
+            are given for free to the player before the beginning of the game.
+
+            'None': No dungeon will be empty. All dungeons may contain useful items
```
Comment (ETR-BTF) [inline SettingsList.py:4070]: Does the option "None" guarantee that the dungeons will have at least 1 item? Or could the randomizer still randomly roll barren dungeons?

File: SettingsList.py [hunk 3]
```diff
@@ -4041,6 +4054,83 @@ def __init__(self, name, gui_text, min, max, default, step=1,
...
+            and have to be beaten in order to get their rewards.
+            'Specific Dungeons': Choose which specific dungeons will be empty.
+            'Count': Choose how many empty dungeons will be randomly chosen.
+
+            Empty dungeons can't be MQ. Incompatible "MQ Dungeon" settings and
```
Comment (ETR-BTF) [inline SettingsList.py:4075]: I would separate the different topics in this paragraph into separate paragraphs.

File: SettingsList.py [hunk 4]
```diff
@@ -3477,6 +3485,11 @@ def __init__(self, name, gui_text, min, max, default, step=1,
             will add 2 more possible locations to each Dungeons.
             This makes dungeons more profitable, especially
             Ice Cavern, Water Temple, and Jabu Jabu's Belly.
+            
+            Regardless of the selected option, maps and compasses from
```
Comment (ETR-BTF) [inline SettingsList.py:3489]: I think the "Empty Dungeons" setting should be mentioned explicitly to avoid confusion with randomly rolled barren dungeons. Also for the (boss) keysanity tooltips.

File: SettingsList.py (hunk 2 above)
Comment (damdayou) [reply SettingsList.py:4070]: No it's a clumsy phrasing from me. 'None' just mean the current way generator works, where some dungeons can be rolled barren. Maybe should I just keep the `No dungeon will be empty.` part and get rid of the rest of the sentence?

File: SettingsList.py (hunk 3 above)
Comment (damdayou) [reply SettingsList.py:4075]: Just to be sure, you mean the paragraph starting line 4075? You think I should split it into four parts:
 - MQ vs empty dungeon,
 - Ignores _Dungeons Have One Major Item_,
 - Dungeon items shuffle,
 - Shuffle songs: dungeon rewards?

File: SettingsList.py (hunk 4 above)
Comment (damdayou) [reply SettingsList.py:3489]: What about starting the paragraph with `If "Empty Dungeons" is on, then regardless of the selected option...`?
Or that also raise a more general question: maybe _Empty dungeons_ is not clear enough and we can find a better wording that leaves no ambiguity with 'naturally' barren dungeons? When @ArthurOudini first presented me this concept, he referred to them as _Banned dungeons_ (having in mind some pick and ban system for 1v1 races for example). That could be an option too.

File: SettingsList.py (hunk 2 above)
Comment (ETR-BTF) [reply SettingsList.py:4070]: "No dungeon will be guaranteed empty and no additional Dungeon Rewards will be given for free, even if dungeons are randomly rolled with no items in them." Something like that maybe?

File: SettingsList.py (hunk 3 above)
Comment (ETR-BTF) [reply SettingsList.py:4075]: Yeah. That makes it easier to read imo.

File: SettingsList.py (hunk 4 above)
Comment (ETR-BTF) [reply SettingsList.py:3489]: I agree that it would probably be best to come up with a phrasing that is less ambiguous.

Comment (ETR-BTF) [reply SettingsList.py:3489]: "Vacant" and "Useless" are also being suggested.

Comment (damdayou) [reply SettingsList.py:3489]: Following discussions on Discord and some more brainstorming, maybe we can narrow down to the names below:
 - vacant dungeon,
 - removed dungeon,
 - cancelled dungeon,
 - nullified dungeon.
 
 Being not fluent in English, I don't know which one describes best _a dungeon filled with junk **and** whose dungeon reward is given for free at the very beginning of the seed_, while not adding confusion with randomly rolled _barren dungeons_. Maybe we can just make a poll on OoTR Discord but I don't know if this is a common way to proceed and I don't want to spam #dev-public-talk.

Once a definitive name is fixed I will commit a tooltip update taking in account all above comments.

**Comment by damdayou** [comment]
Renamed _empty dungeons_ into _pre-completed dungeons_ to avoid confusion with barren dungeons and enforce the idea that the dungeon rewards are automatically collected. GUI text and tooltips have been edited accordingly and taking previous comments into account.

**Comment by cjohnson57** [comment]
A couple things I don't understand:

1) Why is it incompatible with MQ dungeons? You should be able to dynamically pull the location list to fill with junk depending on the MQ setting.

2) Why is it limited to only 6 dungeons?

Furthermore, I think the tooltip should clarify that the setting only applies to blue warp/reward dungeons.

**Comment by damdayou** [comment]
Hey, thanks for taking the time to have a look at this PR.

1. Since empty dungeons are by design not supposed to be visited (except in extremely rare ER seeds), it felt weird for me that a player would specifically ask for them to be MQ. I think that when one asks for say, 3 MQ dungeons and 2 empty dungeons, they really want to play 3 MQ dungeons among the non-empty dungeons. That's why at the moment the code tries to pick independent MQ dungeons and empty dungeons, and raise an error when it's impossible. I can totally change this behavior if needed, and in this case I see two ways of changing it:
 - either I keep trying to pick independent MQ dungeons and empty dungeons whenever it's possible, and when it's not I just allow MQ & empty dungeons instead of raising an error,
 - or I do not make MQ dungeons and empty dungeons independent anymore and when one asks for 3 MQ dungeons and 2 empty dungeons, they can get 2 MQ & empty dungeons + 1 MQ & non-empty dungeon.
 Which of these two options do you prefer?

2. In the early stage of development, asking for more than 6 empty dungeons led to many failed generations because there where too many locked area. But I guess it is very setting-dependent, maybe there are configurations where 8 empty dungeons run perfectly fine. I see no problem raising the limit to 8.

Agreed for the tooltip, I'm going to update it.

**Comment by damdayou** [comment]
I have implemented full compatibility between empty dungeons and MQ dungeons following the first option I suggested (the generator tries as much as possible to have distinct MQ dungeons and empty dungeons, but allows dungeons to be both it needed). If this option is acceptable, I'll just have to edit the tooltip accordingly.

**Comment by cjohnson57** [comment]
Yeah that's probably fine, just didn't want the unnecessary incompatibility/error raising

**Comment by cjohnson57** [comment]
@damdayou It seems good for the most part, however, it seems to error when trying to set songs to dungeon reward locations since the heart locations don't have an associated dungeon, I guess some code needs to run to associate them due to boss ER.
image


Also, it would be nice if you added some unit tests (maybe random/all/half empty dungeons, songs as dungeon rewards on/off, and skip child zelda on/off). If you want some guidance for how to do that, feel free to reach out on the Discord, but some recent PRs such as #1643 are good examples.

**Comment by fenhl** [comment]
To fix the handling of songs on bosses, you can use `hint_area = HintArea.at(location)` to look up where the boss is, check whether `hint_area.is_dungeon` to future-proof for mixed pools bosses, and then check `world.empty_dungeons[hint_area.dungeon_name]`.

It's worth noting that this PR has nontrivial interactions with #1641, so the two should not be merged at the same time. I recommend merging this PR first so I can rebase #1641 and make the required adjustments there.

**Comment by damdayou** [comment]
I have implemented @fenhl's fix to solve the issue with songs on dungeon rewards (thanks again for your advice).
I then added a bunch of unit tests. Maybe that's too much of them, I don't know the standards. Feel free to tell me if I should remove some of them.

**Comment by cjohnson57** [comment]
The unit tests are awesome! Tysm for them. I'll probably merge this later today.

**Review by cjohnson57 — APPROVED** [review]
Since I didn't see it mentioned anywhere, I also like that this PR changes it so starting locations are in their own sphere 0.
